Fix class extraction followed by `(` in Slim

RobinMalfait opened the pull request:
This PR fixes an issue where using the class shorthand in Slim templates, followed by an `(` results in the last class being ignored. 

E.g.:

```slim
body.border-t-4.p-8(class="#{body_classes}" data-hotwire-native="#{hotwire_native_app?}" data-controller="update-time-zone")
```

This is because we will eventually extract `p-8` but it's followed by an invalid boundary character `(`.

To solve this, we make sure to replace the `(` with a space. We already do a similar thing when the classes are followed by an `[`.

One caveat, we _can_ have `(` in our classes, like `bg-(--my-color)`. But in my testing this is not something that can be used in the shorthand version. 

E.g.:
```slim
div.bg-(--my-color)
```

Compiles to:
```html
<div --my-color="" class="bg-"></div>
```

So I didn't add any special handling for this. Even when trying to escape the `(`, `-` and `)` characters, it still doesn't work. E.g.:

```slim
div.bg-\(--my-color\)
```

Compiles to:
```html
<div class="bg-">\(--my-color\)</div>
```

# Test plan

1. Added test for the issue
2. Existing tests pass
3. Verified via the extractor tool:

| Before | After |
| --- | --- |
| <img width="958" alt="image" src="https://github.com/user-attachments/assets/f72c420e-5429-424f-a01d-12f724062bf2" /> | <img width="958" alt="image" src="https://github.com/user-attachments/assets/b0cc8f2f-97a8-4fca-8813-3bb1da8d99a8" /> |

---

Fixes: #17277

RobinMalfait committed:
add failing test

RobinMalfait committed:
ensure class shorthand followed by `(` extracts the last class

RobinMalfait committed:
update changelog

thecrypticace commented on crates/oxide/src/extractor/pre_processors/slim.rs:
```diff
@@ -80,6 +80,23 @@ impl PreProcessor for Slim {
+                    result[cursor.pos] = b' ';
+                    bracket_stack.push(cursor.curr);
+                }
+
```
idk if it matters but should we worry about `bg-red-500/(--my-var)` here too? If so we should do `!matches!(cursor.prev, b'-' | b'/')`

thecrypticace replied:
Okay, answering my own question with a note: I think we can drop that matches check entirely based on the PR desc

This is because the check (after adding the `/`) makes these two tests I just wrote pass:
```rs
let input = r#"
    body.border-(--my-color).p-8(class="\#{body_classes}" data-hotwire-native="\#{hotwire_native_app?}" data-controller="update-time-zone")
"#;
Slim::test_extract_contains(input, vec!["border-(--my-color)", "p-8"]);

let input = r#"
    body.border-red-500/(--my-color).p-8(class="\#{body_classes}" data-hotwire-native="\#{hotwire_native_app?}" data-controller="update-time-zone")
"#;
Slim::test_extract_contains(input, vec!["border-red-500/(--my-color)", "p-8"]);
```

but as you mentioned this is invalid Slim syntax

RobinMalfait replied:
Yep yep, added an additional test just so that this case is covered: https://github.com/tailwindlabs/tailwindcss/pull/17278/commits/1a9c242f4abb37dd15bcda6cb9b451bcb43c9665

RobinMalfait replied:
Alright we _do_ need this for this situation: https://github.com/tailwindlabs/tailwindcss/pull/17278/commits/645db11673316e4bb2e173cca162c5d990c7562e

```slim
div class="bg-(--my-color) bg-(--my-color)/(--my-opacity)"
```

RobinMalfait committed:
add additionaly test with variable shorthand modifier

RobinMalfait committed:
Merge branch 'main' into fix/issue-17277

RobinMalfait committed:
add safety checks around `(`

There is a situation where classes can exist top-level that are _not_
inside of any type of brackets.

E.g.:

```
div class="…"
```
